core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncPathCache.java

```diff
@@ -65,11 +78,62 @@ public boolean shouldSyncPath(String path, long intervalMs) {
       // Always sync.
       return true;
     }
-    Long lastSync = mCache.getIfPresent(path);
-    if (lastSync == null) {
-      // No info about the last sync, so trigger a sync.
-      return true;
+
+    // check the last sync information for the path itself.
+    SyncTime lastSync = mCache.getIfPresent(path);
+    if (lastSync != null && (System.currentTimeMillis() - lastSync.getLastSyncMs()) < intervalMs) {
+      // Sync is not necessary for this path.
+      return false;
+    }
+
+    // sync should be done on this path, but check all ancestors to determine if a recursive sync
+    // had been performed (to avoid a sync again).
+    String currPath = path;
+    while (!currPath.equals(AlluxioURI.SEPARATOR)) {
+      try {
+        currPath = PathUtils.getParent(currPath);
+        lastSync = mCache.getIfPresent(currPath);
+        if (lastSync != null
+            && (System.currentTimeMillis() - lastSync.getLastRecursiveSyncMs()) < intervalMs) {
```

Comment:
because we initialize `syncMs` and `recursiveSyncMs` to -1, it kind of works out here. but could we make it more explicit? like if make this check in a method, and check if `syncMs` or `recursiveSyncMs` is less than 0, we fail the check. because technically `intervalMs` can be set to a very large number, larger than `currentTimeMillis`.

Reply:
Done